database/sqlite3.ksy

```diff
@@ -32,213 +32,431 @@ doc: |
+      the database file locking primitives. SQLite does not use the lock-byte page.
+  ptrmap_page:
+    params:
+      - id: page_number
+        type: u4
+    seq:
+      - id: entries
+        type: ptrmap_entry
+        repeat: expr
+        repeat-expr: entry_count
+    instances:
+      first_page:
+        value: '3 + (_root.header.j * (page_number - 2))'
+      last_page:
+        value: 'first_page + _root.header.j - 1'
+      entry_count:
+        value: '(last_page > _root.header.page_count ? _root.header.page_count : last_page) - first_page + 1'
+  ptrmap_entry:
+    seq:
+      - id: type
+        type: u1
+        enum: ptrmap_page_type
+      - id: page_number
+        type: u4
+  btree_page_pointer:
+    seq:
+      - id: page_number
+        type: u4
+    instances:
+      page:
+        io: _root._io
+        pos: (page_number - 1) * _root.header.page_size
+        size: _root.header.page_size
+        type: btree_page(page_number)
+        if: page_number != 0
   btree_page:
+    params:
+      - id: page_number
+        type: u4
     seq:
+      - id: database_header
+        type: database_header
+        if: page_number == 1
       - id: page_type
         type: u1
+        enum: btree_page_type
       - id: first_freeblock
         type: u2
-      - id: num_cells
+        doc: The start of the first freeblock on the page, or is zero if there are no freeblocks.
+      - id: cell_count
         type: u2
-      - id: ofs_cells
+        doc: The number of cells on the page
+      - id: cell_content_area_start_raw
         type: u2
+        doc: |
+          The start of the cell content area. A zero value for this integer is interpreted as 65536.
+          The interpreted value is available as `cell_content_area`.
       - id: num_frag_free_bytes
         type: u1
+        doc: The number of fragmented free bytes within the cell content area.
       - id: right_ptr
-        type: u4
-        if: page_type == 2 or page_type == 5
+        type: btree_page_pointer
+        if: page_type == btree_page_type::index_interior or page_type == btree_page_type::table_interior
+        doc: |
+          The right-most pointer. This value appears in the header of interior
+          b-tree pages only and is omitted from all other pages.
       - id: cells
-        type: ref_cell
+        type: cell_pointer
         repeat: expr
-        repeat-expr: num_cells
-  ref_cell:
+        repeat-expr: cell_count
+    instances:
+      call_content_area_start:
+        value: 'cell_content_area_start_raw == 0 ? 65536 : cell_content_area_start_raw'
+      cell_content_area:
+        pos: call_content_area_start
+        size: _root.header.usable_size - call_content_area_start
+      reserved_area:
+        pos: _root.header.page_size - _root.header.reserved_space
+        size-eos: true
+        if: _root.header.reserved_space != 0
+  cell_pointer:
```

Comment:
fails to read `cell_pointer` from a large database

```py
    class CellPointer(KaitaiStruct):
        def __init__(self, _io, _parent=None, _root=None):
            self._io = _io
            self._parent = _parent
            self._root = _root if _root else self
            self._read()

        def _read(self):
            self.content_offset = self._io.read_u2be()
```

```
  File "pysqlite3/connection.py", line 228, in _tables
    page = self._db.pages[0] # FIXME this fails on large datbase
  File "pysqlite3/parser/sqlite3.py", line 842, in pages
    Sqlite3.BtreePage((i + 1), _io__raw__m_pages, self, self._root)
  File "pysqlite3/parser/sqlite3.py", line 115, in __init__
    self._read()
  File "pysqlite3/parser/sqlite3.py", line 137, in _read
    self.cells.append(Sqlite3.CellPointer(self._io, self, self._root))
  File "pysqlite3/parser/sqlite3.py", line 697, in __init__
    self._read()
  File "pysqlite3/parser/sqlite3.py", line 700, in _read
    self.content_offset = self._io.read_u2be()
  File "kaitaistruct.py", line 173, in read_u2be
    return KaitaiStream.packer_u2be.unpack(self.read_bytes(2))[0]
  File "kaitaistruct.py", line 326, in read_bytes
    raise EOFError(
EOFError: requested 2 bytes, but only 0 bytes available
```

Reply:
one problem is that `db.pages` is not made for random-access
when i read `db.pages[0]` it will loop through ALL pages

```py
class Sqlite3(KaitaiStruct):
    # ...
    @property
    def pages(self):
        if hasattr(self, "_m_pages"):
            return self._m_pages

        _pos = self._io.pos()
        self._io.seek(0)
        self._raw__m_pages = []
        self._m_pages = []
        for i in range(self.header.page_count):
```

the other problem is that `db.root_pages` is missing
when i follow `right_ptr` then i get a list of all `root_pages`
until `right_ptr.page_number == db.header.page_count`

quickfix for python in https://github.com/milahu/pysqlite3/commit/1256408d5479a4902df17728a2ae6f355fcc44e9

Reply:
I'll have to experiment a bit for this one. Not sure what's going on with the right page pointer. There might simply be an error in the definition.

I currently declared the btree_page_pointer as
```
  btree_page_pointer:
    seq:
      - id: page_number
        type: u4
    instances:
      page:
        io: _root._io
        pos: (page_number - 1) * _root.header.page_size
        size: _root.header.page_size
        type: btree_page(page_number)
        if: page_number != 0
```

What you would actually want to do here is describe that this is a pointer to `db.pages[page_number]`. I'm not sure if there's a way to describe that in Kaitai Struct. I added the `page` instance mainly for my own convenience when browsing through the database file in KSV.

I'm not sure if there's much I can do about the generated python code. The `pages` instances section doesn't really require unmarshalling all database pages up front. I don't know if there's a definition level concept that can inform the Python code generator to produce a lazy collection instead of an eager one.

Reply:
Note that `db.root_pages` doesn't really exist at the level of the database format. You need to query the `sqlite_schema` table (the table b-tree with page 1 as root)  to construct that list if you want it.

Reply:
yes, the parser should be minimal
ie, derive values only when they are needed to continue parsing

ideally kaitai-struct-compiler would have a `debug` option
to generate a more verbose (and more expensive) parse tree

for completeness: 

<details>
<summary>
table_pages vs root_pages
</summary>

> You need to query the `sqlite_schema` table

these are `table_pages`

```py
    def _table_page_numbers(self):
        """
        get the page numbers of all tables,
        excluding the sqlite_schema table in page 1
        """
        page = self._db.root_pages[0]
        assert page.page_type.value == 0x0D  # Table B-Tree Leaf Cell
        for cell in page.cells:
            yield cell.content.payload.values[3].value
```

`root_pages` are more: `table_pages` plus `index_pages` (etc)

```py
    def _root_page_numbers(self):
        """
        get the page numbers of all root pages,
        excluding page 1
        """
        if self._db.header.page_count == 1:
            return []
        i = 1 # page 2
        i_max = self._db.header.page_count - 1
        while i < i_max: # `i == i_max` means "end of file"
            page = self._db.pages[i]
            yield page.page_number
            # go to next root page
            if hasattr(page, "right_ptr"):
                i = page.right_ptr.page_number - 1
            else:
                i = i + 1
```

</details>